flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java

```diff
@@ -0,0 +1,263 @@
+
+    /** scheduled executor, periodically trigger materialization. */
+    private final ScheduledExecutorService periodicExecutor;
+
+    private final CheckpointStreamFactory streamFactory;
+
+    private final Snapshotable keyedStateBackend;
+
+    private final StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter;
+
+    private final AsyncExceptionHandler asyncExceptionHandler;
+
+    private final int allowedNumberOfFailures;
+
+    /** Materialization failure retries. */
+    private final AtomicInteger retries;
+
+    /** Making sure only one materialization on going at a time. */
+    private final AtomicBoolean materializationOnGoing;
+
+    private long materializedId;
+
+    private MaterializedState materializedState;
+
+    public PeriodicMaterializer(
+            MailboxExecutor mailboxExecutor,
+            ExecutorService asyncOperationsThreadPool,
+            Snapshotable keyedStateBackend,
+            CheckpointStorageWorkerView checkpointStorageWorkerView,
+            StateChangelogWriter<ChangelogStateHandle> stateChangelogWriter,
+            AsyncExceptionHandler asyncExceptionHandler,
+            MaterializedState materializedState,
+            long periodicMaterializeInitDelay,
+            long periodicMaterializeInterval,
+            int allowedNumberOfFailures,
+            boolean materializationEnabled) {
+        this.mailboxExecutor = mailboxExecutor;
+        this.asyncOperationsThreadPool = asyncOperationsThreadPool;
+        this.keyedStateBackend = keyedStateBackend;
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.asyncExceptionHandler = asyncExceptionHandler;
+        this.periodicExecutor =
+                Executors.newSingleThreadScheduledExecutor(
+                        new ExecutorThreadFactory("periodic-materialization"));
+        this.streamFactory = shared -> checkpointStorageWorkerView.createTaskOwnedStateStream();
+        this.allowedNumberOfFailures = allowedNumberOfFailures;
+        this.materializationOnGoing = new AtomicBoolean(false);
+        this.retries = new AtomicInteger(allowedNumberOfFailures);
+
+        this.materializedId = 0;
+        this.materializedState = materializedState;
+
+        if (materializationEnabled) {
+            this.periodicExecutor.scheduleAtFixedRate(
+                    this::triggerMaterialization,
+                    periodicMaterializeInitDelay,
+                    periodicMaterializeInterval,
+                    TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // Only one materialization ongoing at a time
+                    if (!materializationOnGoing.compareAndSet(false, true)) {
+                        return;
+                    }
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
```

Comment:
NON-materialized is not part of the materialized state, that part I agree.
However, I do think NON-materialized + materialized state needs to be put together.

How about this: put NON-materialized, materialized state and seq number together as ChangelogKeyedStateBackend's state, and provides API to access this state from PeriodicMaterializer.

Reply:
The reason is that I still want to keep all the materialization (sync and async phase within the materializer), and do not want to expose the details.

The only interaction between Keyedstatebackend and materializer is Keyedstatebackend `state` update

Reply:
> How about this: put NON-materialized, materialized state and seq number together as ChangelogKeyedStateBackend's state, and provides API to access this state from PeriodicMaterializer.

This is similar to what I proposed above, except that you want materializer, not the backend run the sync phase, right?
I guess the motivation is to have "materialization" in code covering both sync and async phases?

This may be reasonable, but OTH we lose binding between SQN and sync-snapshot which makes it more error-prone.
Besides that, the final solution seems cleaner if sync-phase runs in backend, as all mailbox-requiring activities may reside in one class.

Reply:
1. the sqn and sync-snapshot are still bonded together in my current solution

2. I want to sync + async to be wrapped in the materializer; the backend is only responsible to provide a way to make a snapshot.

Changelog statebackend contains logic -> materilizer + changed log 

materializer -> sync + async

ideally, I think changed log is better to be wrapped as well, but I am fine with it right now.

Reply:
Resolved by using the proposal [above](https://github.com/apache/flink/pull/16606#discussion_r682621905).